atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -472,16 +474,19 @@ public Map<byte[], Iterator<Map.Entry<Cell, byte[]>>> getRowsColumnRangeIterator
                 keyValueService.getRowsColumnRange(tableRef, rows, columnRangeSelection, getStartTimestamp());
         ImmutableSortedMap.Builder<byte[], Iterator<Map.Entry<Cell, byte[]>>> postFilteredResultsBuilder =
                 ImmutableSortedMap.orderedBy(PtBytes.BYTES_COMPARATOR);
-        for (Map.Entry<byte[], RowColumnRangeIterator> e : rawResults.entrySet()) {
-            byte[] row = e.getKey();
-            RowColumnRangeIterator rawIterator = e.getValue();
-
-            // Sadly we can't close properly here without an ABI break :/
-            @SuppressWarnings("MustBeClosedChecker")
-            ClosableIterator<Map.Entry<Cell, byte[]>> postFilteredIterator =
-                    getPostFilteredColumns(tableRef, columnRangeSelection, row, rawIterator);
-
-            postFilteredResultsBuilder.put(row, scopeToTransaction(postFilteredIterator));
+        for (byte[] row : rows) {
+            Iterator<Entry<Cell, byte[]>> entryIterator = Optional.ofNullable(rawResults.get(row))
+                    .<Iterator<Entry<Cell, byte[]>>>map(rawIterator -> {
+                        // Sadly we can't close properly here without an ABI break :/
+                        @SuppressWarnings("MustBeClosedChecker")
+                        ClosableIterator<Entry<Cell, byte[]>> postFilteredIterator =
+                                getPostFilteredColumns(tableRef, columnRangeSelection, row, rawIterator);
+                        return postFilteredIterator;
+                    })
+                    .orElseGet(() -> getLocalWritesForColumnRange(tableRef, columnRangeSelection, row)
+                            .entrySet()
+                            .iterator());
+            postFilteredResultsBuilder.put(row, scopeToTransaction(entryIterator));
         }
         SortedMap<byte[], Iterator<Map.Entry<Cell, byte[]>>> postFilteredResults =
                 postFilteredResultsBuilder.buildOrThrow();
```

Comment:
style nit: could stream collect into the  here:

```suggestion
        ImmutableSortedMap<byte[], Iterator<Entry<Cell, byte[]>>> postFilteredResults = Streams.stream(rows)
                .collect(ImmutableSortedMap.toImmutableSortedMap(PtBytes.BYTES_COMPARATOR, row -> row, row -> {
                    // explicitly not using Optional due to allocation perf overhead
                    Iterator<Entry<Cell, byte[]>> entryIterator;
                    RowColumnRangeIterator rawIterator = rawResults.get(row);
                    if (rawIterator != null) {
                        // Sadly we can't close properly here without an ABI break :/
                        entryIterator = getPostFilteredColumns(tableRef, columnRangeSelection, row, rawIterator);
                    } else {
                        entryIterator = getLocalWritesForColumnRange(tableRef, columnRangeSelection, row)
                                .entrySet()
                                .iterator();
                    }
                    return scopeToTransaction(entryIterator);
                }));
```